### PR TITLE
Make HLists covariant in their type parameters.

### PR DESCRIPTION
**Pull request by Shadowfiend**
Also add a spec for HList length.

I'm using this code in lift-formality and it works as advertised, but it's nicer if
I don't keep my own copy of it ;)

### CODE REVIEW

File: core/common/src/main/scala/net/liftweb/common/HList.scala [hunk 1]
```diff
@@ -116,7 +96,7 @@ object HLists {
    * Carries the information about the type of this element, plus the `HList`
    * type of the rest of the list.
    *
-   * You can use `:+:` to make this `HList` longer:
+   * You can use `[[HListMethods.:+: :+:]]` to make this `HList` longer:
    *
    * {{{
```
Comment (fmpwizard) [inline core/common/src/main/scala/net/liftweb/common/HList.scala]: would you want to update the scaladoc style while you are here or are you planning on doing another massive pass at docs on a separate PR?

**Comment by fmpwizard** [comment]
:+1:

File: core/common/src/main/scala/net/liftweb/common/HList.scala (hunk 1 above)
Comment (Shadowfiend) [inline core/common/src/main/scala/net/liftweb/common/HList.scala]: I actually waited until the other docs PR for `lift-common` went in, so these docs should all be up-to-date, I think. Did you see an issue? I kind of had to move the tweaks I'd made for lift-formality into the already-properly-doced file, so it's possible I missed something.

**Comment by Shadowfiend** [comment]
I'm not sure how much HLists are being used in general, but given one thumbs up I'll probably merge this tomorrow unless someone else protests vehemently :p

**Comment by Shadowfiend** [comment]
I'm not sure when “tomorrow” turned into “in 6 days”, but in we go :p

**Pull request merged**